LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DownlinkMessageBuilder.cs

```diff
@@ -202,27 +200,50 @@ internal static class DownlinkMessageBuilder
                 1,
                 loRaDevice.Supports32BitFCnt ? fcntDown : null);
 
+            // following calculation is making sure that ReportedRXDelay is chosen if not default,
+            // otherwise for class C devices a LnsRxDelay of 0 is chosen, for class A devices a value of 1 is chosen
+            var lnsRxDelay = Math.Max((loRaDevice.ClassType is LoRaDeviceClassType.C ? RxDelay0 : RxDelay1).ToSeconds(), loRaDevice.ReportedRXDelay.ToSeconds());
+
             // todo: check the device twin preference if using confirmed or unconfirmed down
-            var downlinkMessage = BuildDownstreamMessage(loRaDevice, request.StationEui, logger, radioMetadata.UpInfo.Xtime, datr, loRaRegion.GetDownstreamRX2DataRate(configuration.Rx2DataRate, loRaDevice.ReportedRX2DataRate, logger, deviceJoinInfo), freq, loRaRegion.GetDownstreamRX2Freq(configuration.Rx2Frequency, logger, deviceJoinInfo), lnsRxDelay, ackLoRaMessage, radioMetadata.UpInfo.AntennaPreference);
+            var downlinkMessage = BuildDownstreamMessage(loRaDevice,
+                                                         request.StationEui,
+                                                         logger,
+                                                         radioMetadata.UpInfo.Xtime,
+                                                         receiveWindow == Constants.ReceiveWindow2 ? null : (datr, freq),
+                                                         rx2,
+                                                         (RxDelay)lnsRxDelay,
+                                                         ackLoRaMessage,
+                                                         loRaDevice.ClassType,
+                                                         radioMetadata.UpInfo.AntennaPreference);
 
             if (logger.IsEnabled(LogLevel.Debug))
                 logger.LogDebug($"{ackLoRaMessage.MessageType} {JsonConvert.SerializeObject(downlinkMessage)}");
 
             return new DownlinkMessageBuilderResponse(downlinkMessage, isMessageTooLong, receiveWindow);
         }
 
-        private static DownlinkMessage BuildDownstreamMessage(LoRaDevice loRaDevice, StationEui stationEUI, ILogger logger, ulong xTime, DataRateIndex rx1Datr, DataRateIndex rx2Datr, Hertz freqRx1, Hertz freqRx2, ushort lnsRxDelay, LoRaPayloadData loRaMessage, uint? antennaPreference = null)
+        private static DownlinkMessage BuildDownstreamMessage(LoRaDevice loRaDevice,
+                                                              StationEui stationEUI,
+                                                              ILogger logger,
+                                                              ulong xTime,
+                                                              (DataRateIndex datr, Hertz freq)? rx1,
```

Comment:
nit: You could skip the tuple element names here and deconstruct in the method:

```c#
var (rx1DataRateIndex, rx1Freq) = rx1.GetValueOrDefault();
var (rx2DataRateIndex, rx2Freq) = rx2;
```

This way, `GetValueOrDefault` is also only called once.